File: goldencheck/profilers/type_inference.py
"""Type inference profiler — detects mixed types and type mismatches."""
from __future__ import annotations
import polars as pl
from goldencheck.models.finding import Finding, Severity
from goldencheck.profilers.base import BaseProfiler

class TypeInferenceProfiler(BaseProfiler):
    def profile(self, df: pl.DataFrame, column: str, *, context: dict | None = None) -> list[Finding]:
        findings: list[Finding] = []
        col = df[column]
        dtype = col.dtype
        if dtype == pl.Utf8 or dtype == pl.String:
            non_null = col.drop_nulls()
            if len(non_null) == 0:
                return findings
            cast_result = non_null.cast(pl.Float64, strict=False)
            numeric_count = cast_result.is_not_null().sum()
            numeric_pct = numeric_count / len(non_null) if len(non_null) > 0 else 0
            if numeric_pct >= 0.8:
                int_cast = non_null.cast(pl.Int64, strict=False)
                int_count = int_cast.is_not_null().sum()
                int_pct = int_count / len(non_null)
                type_name = "integer" if int_pct > 0.9 else "numeric"
                non_numeric = len(non_null) - numeric_count
                # Write context so other profilers can chain
                if context is not None:
                    context.setdefault(column, {})["mostly_numeric"] = True
                findings.append(Finding(
                    severity=Severity.WARNING, column=column, check="type_inference",
                    message=f"Column is string but {numeric_pct:.0%} of values are {type_name} ({non_numeric} non-{type_name} values)",
                    affected_rows=non_numeric,
                    suggestion=f"Consider casting to {type_name}",
                    confidence=0.9,
                ))
            elif numeric_pct > 0 and numeric_pct < 0.05:
                # Minority numeric values in a mostly-text column — suspicious but low confidence
                minority_count = numeric_count
                findings.append(Finding(
                    severity=Severity.INFO, column=column, check="type_inference",
                    message=f"Column is string but {numeric_pct:.1%} of values appear numeric ({minority_count} values) — possible data entry error",
                    affected_rows=minority_count,
                    suggestion="Investigate numeric values in this text column",
                    confidence=0.3,
                ))
        return findings

File: packages/goldencheck-js/src/core/profilers/type-inference.ts
/**
 * Type inference profiler — detects mixed types and type mismatches.
 * Port of goldencheck/profilers/type_inference.py.
 */

import type { TabularData } from "../data.js";
import { isNullish } from "../data.js";
import { type Finding, Severity, makeFinding } from "../types.js";
import type { Profiler } from "./base.js";

const SHOULD_BE_STRING = ["zip", "postal", "phone", "fax", "ssn", "npi", "id", "code", "sku"];

export class TypeInferenceProfiler implements Profiler {
  profile(
    data: TabularData,
    column: string,
    context?: Record<string, unknown>,
  ): Finding[] {
    const findings: Finding[] = [];
    const values = data.column(column);
    const nonNull = values.filter((v) => !isNullish(v));
    const total = nonNull.length;
    if (total === 0) return findings;

    // Detect string column that is mostly numeric
    const dt = data.dtype(column);
    if (dt === "string") {
      let numericCount = 0;
      let intCount = 0;
      for (const v of nonNull) {
        const n = Number(v);
        if (Number.isFinite(n)) {
          numericCount++;
          if (Number.isInteger(n)) intCount++;
        }
      }
      const numericPct = numericCount / total;

      if (numericPct >= 0.8) {
        const intPct = intCount / total;
        const typeName = intPct > 0.9 ? "integer" : "numeric";
        const nonNumeric = total - numericCount;

        if (context) {
          if (!context[column]) context[column] = {};
          (context[column] as Record<string, unknown>)["mostly_numeric"] = true;
        }

        findings.push(
          makeFinding({
            severity: Severity.WARNING,
            column,
            check: "type_inference",
            message: `Column is string but ${Math.round(numericPct * 100)}% of values are ${typeName} (${nonNumeric} non-${typeName} values)`,
            affectedRows: nonNumeric,
            suggestion: `Consider casting to ${typeName}`,
            confidence: 0.9,
          }),
        );
      } else if (numericPct > 0 && numericPct < 0.05) {
        findings.push(
          makeFinding({
            severity: Severity.INFO,
            column,
            check: "type_inference",
            message: `Column is string but ${(numericPct * 100).toFixed(1)}% of values appear numeric (${numericCount} values) — possible data entry error`,
            affectedRows: numericCount,
            suggestion: "Investigate numeric values in this text column",
            confidence: 0.3,
          }),
        );
      }
    }

    // Check: numeric column that should be string based on name
    if (dt === "integer" || dt === "float") {
      const colLower = column.toLowerCase();
      for (const hint of SHOULD_BE_STRING) {
        if (colLower.includes(hint)) {
          findings.push(
            makeFinding({
              severity: Severity.WARNING,
              column,
              check: "type_inference",
              message: `Column '${column}' is numeric but name suggests it should be string (may lose leading zeros)`,
              confidence: 0.6,
              suggestion: "Consider storing as string to preserve formatting",
            }),
          );
          break;
        }
      }
    }

    return findings;
  }
}
